File: src/tensorflow/ops/io/rpc.nim
import ../../core

{.compile:"../../ops/io/rpc/rpc.cc".}


type Rpc* {.header:"../../ops/io/rpc/rpc.h", importcpp:"Rpc/*'0*/".} = object
  operation*: Operation[ostring]
  output*: ostring

proc iirpc(scope: Scope,
          address: ostring,
          nmethod: ostring,
          request: ostring,
          protocol: cstring,
          fail_fast: bool,
          timeout_in_ms: int64): Rpc {.header:"../../ops/io/rpc/rpc.h", importcpp:"Rpc(*#, #, #, #, tensorflow::string(#), #, #)", constructor.}

proc rpc*(scope: Scope,
          address: ostring,
          nmethod: ostring,
          request: ostring,
          protocol: cstring,
          fail_fast: bool = true,
          timeout_in_ms: int64 = 0.int): Rpc =
  return iirpc(scope,
               address,
               nmethod,
               request,
               protocol,
               fail_fast,
               timeout_in_ms)

converter rpcToOut*(op: Rpc): ostring {.inline.} = return op.output


type TryRpc* {.header:"../../ops/io/rpc/rpc.h", importcpp:"TryRpc/*'0*/".} = object
  operation*: Operation[ostring]
  output*: ostring

proc iitryRpc(scope: Scope,
             address: ostring,
             nmethod: ostring,
             request: ostring,
             protocol: cstring,
             fail_fast: bool,
             timeout_in_ms: int64): TryRpc {.header:"../../ops/io/rpc/rpc.h", importcpp:"TryRpc(*#, #, #, #, tensorflow::string(#), #, #)", constructor.}

proc tryRpc*(scope: Scope,
             address: ostring,
             nmethod: ostring,
             request: ostring,
             protocol: cstring,
             fail_fast: bool = true,
             timeout_in_ms: int64 = 0.int): TryRpc =
  return iitryRpc(scope,
                  address,
                  nmethod,
                  request,
                  protocol,
                  fail_fast,
                  timeout_in_ms)

converter tryRpcToOut*(op: TryRpc): ostring {.inline.} = return op.output

File: src/tensorflow/ops/io/rpc/rpc.cc
#include "tensorflow/cc/ops/const_op.h"
#include "../../ops/io/rpc/rpc.h"

Rpc::Rpc(tensorflow::Scope& scope, 
           tensorflow::Input address, 
           tensorflow::Input method, 
           tensorflow::Input request, 
           tensorflow::string protocol, 
           bool fail_fast, 
           int64_t timeout_in_ms) {
      if (!scope.ok())
          return;
      auto _address = ::tensorflow::ops::AsNodeOut(scope, address);
      if (!scope.ok())
          return;
      auto _method = ::tensorflow::ops::AsNodeOut(scope, method);
      if (!scope.ok())
          return;
      auto _request = ::tensorflow::ops::AsNodeOut(scope, request);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("Rpc");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "Rpc")
                                   .Input(_address)
                                   .Input(_method)
                                   .Input(_request)
                                   .Attr("protocol", protocol)
                                   .Attr("fail_fast", fail_fast)
                                   .Attr("timeout_in_ms", timeout_in_ms)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      this->output = tensorflow::Output(ret, 0);
}

TryRpc::TryRpc(tensorflow::Scope& scope, 
           tensorflow::Input address, 
           tensorflow::Input method, 
           tensorflow::Input request, 
           tensorflow::string protocol, 
           bool fail_fast, 
           int64_t timeout_in_ms) {
      if (!scope.ok())
          return;
      auto _address = ::tensorflow::ops::AsNodeOut(scope, address);
      if (!scope.ok())
          return;
      auto _method = ::tensorflow::ops::AsNodeOut(scope, method);
      if (!scope.ok())
          return;
      auto _request = ::tensorflow::ops::AsNodeOut(scope, request);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("TryRpc");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "TryRpc")
                                   .Input(_address)
                                   .Input(_method)
                                   .Input(_request)
                                   .Attr("protocol", protocol)
                                   .Attr("fail_fast", fail_fast)
                                   .Attr("timeout_in_ms", timeout_in_ms)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      this->output = tensorflow::Output(ret, 0);
}

